# Data Cleaning Challenge

## By Chinelo Cynthia Ezenwafor




## Table of Contents

* [Introduction](#Introduction) 

* [Data Assessment](#Data_Assessment) 

* [Data Cleaning](#Data_Cleaning)  

<a id = 'Introduction'></a>
## Introduction

> The data Cleaning Challenge was organised on twitter by Promise Nonso and Victor Somadina in order to help data analysts train/improve on their data cleaning skills using any tool of their choice. It also provided an avenue for data analysts to meet and collaborate with other learners. The dataset for this challenge was sourced from kaggle and can be found [here](https://www.kaggle.com/datasets/yagunnersya/fifa-21-messy-raw-dataset-for-cleaning-exploring). 

In [1]:
# import all packages and set plots to be embedded inline
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

%matplotlib inline
pd.set_option('display.max_columns', None)

In [2]:
# read the dataset into a dataframe
fifa = pd.read_csv('FIFA_2021.csv', low_memory=False)

In [3]:
print(fifa.shape)
fifa.head()

(18979, 77)


ID               Name                      LongName  \
0  158023           L. Messi                  Lionel Messi   
1   20801  Cristiano Ronaldo  C. Ronaldo dos Santos Aveiro   
2  200389           J. Oblak                     Jan Oblak   
3  192985       K. De Bruyne               Kevin De Bruyne   
4  190871          Neymar Jr    Neymar da Silva Santos Jr.   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   
1  https://cdn.sofifa.com/players/020/801/21_60.png   
2  https://cdn.sofifa.com/players/200/389/21_60.png   
3  https://cdn.sofifa.com/players/192/985/21_60.png   
4  https://cdn.sofifa.com/players/190/871/21_60.png   

                                           playerUrl Nationality  Age  OVA  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   93   
1  http://sofifa.com/player/20801/c-ronaldo-dos-s...    Portugal   35   92   
2  http://sofifa.com/player/200389/jan-oblak/210006/    Slovenia   27   91   
3  http://sofifa.com/player/192985/kevin-de-bruyn...     Belgium   29   91   
4  http://sofifa.com/player/190871/neymar-da-silv...      Brazil   28   91   

   POT                         Club     Contract   Positions Height Weight  \
0   93         \n\n\n\nFC Barcelona  2004 ~ 2021  RW, ST, CF  170cm   72kg   
1   92             \n\n\n\nJuventus  2018 ~ 2022      ST, LW  187cm   83kg   
2   93      \n\n\n\nAtlético Madrid  2014 ~ 2023          GK  188cm   87kg   
3   91      \n\n\n\nManchester City  2015 ~ 2023     CAM, CM  181cm   70kg   
4   91  \n\n\n\nParis Saint-Germain  2017 ~ 2022     LW, CAM  175cm   68kg   

  Preferred Foot  BOV Best Position     Joined Loan Date End    Value   Wage  \
0           Left   93            RW   1-Jul-04           NaN  €103.5M  €560K   
1          Right   92            ST  10-Jul-18           NaN     €63M  €220K   
2          Right   91            GK  16-Jul-14           NaN    €120M  €125K   
3          Right   91           CAM  30-Aug-15           NaN    €129M  €370K   
4          Right   91            LW   3-Aug-17           NaN    €132M  €270K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   
1         €75.9M        437        84         95                90   
2        €159.4M         95        13         11                15   
3          €161M        407        94         82                55   
4        €166.5M        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \

> Dataset contains 77 details about 18,979 players.

Variable description
Here is a brief documentation for each column name in the given dataset:

- photoUrl: The URL of the player's photo.
- LongName: The full name of the player.
- playerUrl: The URL of the player's page on sofifa.com.
- Nationality: The nationality of the player.
- Positions: The positions the player can play.
- Name: The short name of the player.
- Age: The age of the player.
- OVA: The overall rating of the player in FIFA 21.
- POT: The potential rating of the player in FIFA 21.
- Team & Contract: The team the player is playing for in FIFA 21, along with their contract details.
- ID: The unique identifier for the player.
- Height: The height of the player in feet and inches.
- Weight: The weight of the player in pounds.
- foot: The preferred foot of the player.
- BOV: The best overall rating the player has achieved in their career.
- BP: The best position the player has played in their career.
- Growth: The difference between the potential rating and overall rating of the player.
- Joined: The date the player joined their current team in FIFA 21.
- Loan Date End: The date the player's loan contract ends.
- Value: The market value of the player in FIFA 21.
- Wage: The weekly wage of the player in FIFA 21.
- Release Clause: The release clause value of the player in FIFA 21.
- Attacking: The attacking attributes of the player.
- Crossing: The crossing attribute of the player.
- Finishing: The finishing attribute of the player.
- Heading Accuracy: The heading accuracy attribute of the player.
- Short Passing: The short passing attribute of the player.
- Volleys: The volleys attribute of the player.
- Skill: The skill attributes of the player.
- Dribbling: The dribbling attribute of the player.
- Curve: The curve attribute of the player.
- FK Accuracy: The free kick accuracy attribute of the player.
- Long Passing: The long passing attribute of the player.
- Ball Control: The ball control attribute of the player.
- Movement: The movement attributes of the player.
- Acceleration: The acceleration attribute of the player.
- Sprint Speed: The sprint speed attribute of the player.
- Agility: The agility attribute of the player.
- Reactions: The reactions attribute of the player.
- Balance: The balance attribute of the player.
- Power: The power attributes of the player.
- Shot Power: The shot power attribute of the player.
- Jumping: The jumping attribute of the player.
- Stamina: The stamina attribute of the player.
- Strength: The strength attribute of the player.
- Long Shots: The long shots attribute of the player.
- Mentality: The mentality attributes of the player.
- Aggression: The aggression attribute of the player.
- Interceptions: The interceptions attribute of the player.
- Positioning: The positioning attribute of the player.
- Vision: The vision attribute of the player.
- Penalties: The penalties attribute of the player.
- Composure: The composure attribute of the player.
- Defending: The defending attributes of the player.
- Marking: The marking attribute of the player.
- Standing Tackle: The standing tackle attribute of the player.
- Sliding Tackle: The sliding tackle attribute of the player.
- Goalkeeping: The goalkeeping attributes of the player.
- GK Diving: The goalkeeper diving attribute of the player.
- GK Handling: The goalkeeper handling attribute of the player.
- GK Kicking: The goalkeeper kicking attribute of the player.
- GK Positioning: The goalkeeper positioning attribute of the player.
- GK Reflexes: This refers to the goalkeeper's ability to react and make saves quickly.
- Total Stats: This refers to the overall rating of the player based on their performance in all areas of the game.
- Base Stats: This refers to the player's rating in the six main areas of the game: Pace, Shooting, Passing, Dribbling,   Defending, and Physicality.
- W/F: This refers to the player's weaker foot ability.
- SM: This refers to the player's skill moves ability.
- A/W: This refers to the player's attacking work rate. It measures how frequently the player participates in attacking actions, such as making runs or positioning themselves in the opponent's half.
- D/W: This refers to the player's defensive work rate. It measures how frequently the player participates in defensive actions, such as tracking back or making tackles.
- IR: This refers to the player's injury resistance. It measures the player's ability to avoid injuries and how quickly they recover from them.
- PAC: This refers to the player's pace or speed attribute. It measures how quickly the player can move with and without the ball.
- SHO: This refers to the player's shooting ability. It measures the player's accuracy and power when shooting the ball.
- PAS: This refers to the player's passing ability. It measures the player's accuracy and range when passing the ball.
- DRI: This refers to the player's dribbling ability. It measures the player's agility, balance, and ball control when dribbling the ball.
- DEF: This refers to the player's defensive ability. It measures the player's ability to tackle, intercept, and defend against opposing players.
- PHY: This refers to the player's physicality or strength. It measures the player's ability to win physical battles and maintain possession of the ball.
- Hits: This refers to the number of times the player's profile has been viewed on the website.

<a id = 'Data_Assessment'></a>
## Data Assessment

Here, we assess our dataset for findings on quality and tidiness issues.

In [4]:
# statistically analyse the dataset
fifa.describe()

ID           Age           OVA           POT           BOV  \
count   18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean   226403.384794     25.194109     65.718636     71.136414     66.751726   
std     27141.054157      4.710520      6.968999      6.114635      6.747193   
min        41.000000     16.000000     47.000000     47.000000     48.000000   
25%    210135.000000     21.000000     61.000000     67.000000     62.000000   
50%    232418.000000     25.000000     66.000000     71.000000     67.000000   
75%    246922.500000     29.000000     70.000000     75.000000     71.000000   
max    259216.000000     53.000000     93.000000     95.000000     93.000000   

          Attacking      Crossing     Finishing  Heading Accuracy  \
count  18979.000000  18979.000000  18979.000000      18979.000000   
mean     248.938142     49.688392     45.842405         51.942726   
std       74.299428     18.131153     19.567081         17.294409   
min       42.000000      6.000000      3.000000          5.000000   
25%      222.000000     38.000000     30.000000         44.000000   
50%      263.000000     54.000000     49.000000         55.000000   
75%      297.000000     63.000000     62.000000         64.000000   
max      437.000000     94.000000     95.000000         93.000000   

       Short Passing       Volleys         Skill     Dribbling         Curve  \
count   18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean       58.768112     42.696507    256.479214     55.587491     47.246957   
std        14.519106     17.646937     78.650601     18.761314     18.207790   
min         7.000000      3.000000     40.000000      5.000000      4.000000   
25%        54.000000     30.000000    222.000000     49.000000     35.000000   
50%        62.000000     44.000000    269.000000     61.000000     49.000000   
75%        68.000000     56.000000    310.000000     68.000000     61.000000   
max        94.000000     90.000000    470.000000     96.000000     94.000000   

        FK Accuracy  Long Passing  Ball Control      Movement  Acceleration  \
count  18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean      42.391011     52.725381     58.528373    317.718689     64.357553   
std       17.227947     15.178151     16.565892     55.879046     14.889981   
min        5.000000      5.000000      5.000000    122.000000     13.000000   
25%       31.000000     43.000000     54.000000    289.000000     57.000000   
50%       41.000000     56.000000     63.000000    327.000000     67.000000   
75%       55.000000     64.000000     69.000000    356.000000     74.000000   
max       94.000000     93.000000     96.000000    464.000000     97.000000   

       Sprint Speed       Agility     Reactions       Balance         Power  \
count  18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean      64.412930     63.366721     61.655619     63.925865    296.605195   
std       14.638743     14.596277      9.072114     14.062285     50.732390   
min       12.000000     14.000000     24.000000     12.000000    122.000000   
25%       57.000000     55.000000     56.000000     56.000000    264.000000   
50%       67.000000     66.000000     62.000000     66.000000    302.000000   
75%       74.000000     74.000000     68.000000     74.000000    334.000000   
max       96.000000     96.000000     95.000000     97.000000    444.000000   

         Shot Power       Jumping       Stamina      Strength    Long Shots  \
count  18979.000000  18979.000000  18979.000000  18979.000000  18979.000000   
mean      57.801676     64.577375     62.653986     64.786922     46.785236   
std       13.308747     11.844870     15.804223     12.488672     19.300534   
min       18.000000     15.000000     12.000000     16.000000      4.000000   
25%       48.000000     58.000000     55.000000     57.000000     32.000000   
50%       59.000000     65.000000     66.000000     66.0

In [5]:
# to get a concise summary of the dataset
fifa.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 18979 entries, 0 to 18978
Data columns (total 77 columns):
 #   Column            Non-Null Count  Dtype 
---  ------            --------------  ----- 
 0   ID                18979 non-null  int64 
 1   Name              18979 non-null  object
 2   LongName          18979 non-null  object
 3   photoUrl          18979 non-null  object
 4   playerUrl         18979 non-null  object
 5   Nationality       18979 non-null  object
 6   Age               18979 non-null  int64 
 7   OVA               18979 non-null  int64 
 8   POT               18979 non-null  int64 
 9   Club              18979 non-null  object
 10  Contract          18979 non-null  object
 11  Positions         18979 non-null  object
 12  Height            18979 non-null  object
 13  Weight            18979 non-null  object
 14  Preferred Foot    18979 non-null  object
 15  BOV               18979 non-null  int64 
 16  Best Position     18979 non-null  object
 17  Joined      

In [6]:
#check for duplicates
sum(fifa.duplicated())

0

> Dataset contains no duplicates.

In [7]:
pd.set_option('display.max_rows', None)
# assess the dataframe for columns with missing values
fifa.isna().sum()

ID                      0
Name                    0
LongName                0
photoUrl                0
playerUrl               0
Nationality             0
Age                     0
OVA                     0
POT                     0
Club                    0
Contract                0
Positions               0
Height                  0
Weight                  0
Preferred Foot          0
BOV                     0
Best Position           0
Joined                  0
Loan Date End       17966
Value                   0
Wage                    0
Release Clause          0
Attacking               0
Crossing                0
Finishing               0
Heading Accuracy        0
Short Passing           0
Volleys                 0
Skill                   0
Dribbling               0
Curve                   0
FK Accuracy             0
Long Passing            0
Ball Control            0
Movement                0
Acceleration            0
Sprint Speed            0
Agility                 0
Reactions   

Only the columns; Loan Date End and Hits contain missing/null values

In [8]:
# Going through the unique values in the height column
fifa['Height'].unique()

array(['170cm', '187cm', '188cm', '181cm', '175cm', '184cm', '191cm',
       '178cm', '193cm', '185cm', '199cm', '173cm', '168cm', '176cm',
       '177cm', '183cm', '180cm', '189cm', '179cm', '195cm', '172cm',
       '182cm', '186cm', '192cm', '165cm', '194cm', '167cm', '196cm',
       '163cm', '190cm', '174cm', '169cm', '171cm', '197cm', '200cm',
       '166cm', '6\'2"', '164cm', '198cm', '6\'3"', '6\'5"', '5\'11"',
       '6\'4"', '6\'1"', '6\'0"', '5\'10"', '5\'9"', '5\'6"', '5\'7"',
       '5\'4"', '201cm', '158cm', '162cm', '161cm', '160cm', '203cm',
       '157cm', '156cm', '202cm', '159cm', '206cm', '155cm'], dtype=object)

In [9]:
# Going through the unique values in the weight column
fifa['Weight'].unique()

array(['72kg', '83kg', '87kg', '70kg', '68kg', '80kg', '71kg', '91kg',
       '73kg', '85kg', '92kg', '69kg', '84kg', '96kg', '81kg', '82kg',
       '75kg', '86kg', '89kg', '74kg', '76kg', '64kg', '78kg', '90kg',
       '66kg', '60kg', '94kg', '79kg', '67kg', '65kg', '59kg', '61kg',
       '93kg', '88kg', '97kg', '77kg', '62kg', '63kg', '95kg', '100kg',
       '58kg', '183lbs', '179lbs', '172lbs', '196lbs', '176lbs', '185lbs',
       '170lbs', '203lbs', '168lbs', '161lbs', '146lbs', '130lbs',
       '190lbs', '174lbs', '148lbs', '165lbs', '159lbs', '192lbs',
       '181lbs', '139lbs', '154lbs', '157lbs', '163lbs', '98kg', '103kg',
       '99kg', '102kg', '56kg', '101kg', '57kg', '55kg', '104kg', '107kg',
       '110kg', '53kg', '50kg', '54kg', '52kg'], dtype=object)

In [10]:
# Going through the unique values in the weight column
fifa['Wage'].unique()

array(['€560K', '€220K', '€125K', '€370K', '€270K', '€240K', '€250K',
       '€160K', '€260K', '€210K', '€310K', '€130K', '€350K', '€300K',
       '€190K', '€145K', '€195K', '€100K', '€140K', '€290K', '€82K',
       '€110K', '€230K', '€155K', '€200K', '€165K', '€95K', '€170K',
       '€105K', '€115K', '€150K', '€135K', '€55K', '€58K', '€81K', '€34K',
       '€120K', '€59K', '€90K', '€65K', '€56K', '€71K', '€18K', '€75K',
       '€47K', '€20K', '€84K', '€86K', '€74K', '€78K', '€27K', '€68K',
       '€85K', '€25K', '€46K', '€83K', '€54K', '€79K', '€175K', '€43K',
       '€49K', '€45K', '€38K', '€41K', '€39K', '€23K', '€51K', '€50K',
       '€87K', '€30K', '€14K', '€69K', '€31K', '€64K', '€53K', '€35K',
       '€21K', '€28K', '€17K', '€33K', '€70K', '€32K', '€89K', '€26K',
       '€40K', '€76K', '€72K', '€48K', '€36K', '€29K', '€60K', '€16K',
       '€37K', '€24K', '€52K', '€0', '€62K', '€73K', '€63K', '€19K',
       '€1K', '€66K', '€80K', '€12K', '€2K', '€42K', '€13K', '€900',
       '€5

In [11]:
# Going through the unique values in the weight column
fifa['Release Clause'].unique()

array(['€138.4M', '€75.9M', '€159.4M', ..., '€59K', '€35K', '€64K'],
      dtype=object)

### Assessment summary

The following issues were detected after programmatically and visually assessing the dataframe;

> 1. Many of the columns such as height, weight, value, wage etc were wrongly stored as string data type instead of integers, and this is because of the presence of alphabetical characters.
> 2. Some of the column names are not clear such POT, OVA etc and need to be renamed to enhance readability.
> 3. Some columns are not necessary/needed for analysis hence they should be dropped.
> 4. The club column contains '/n/n/n' which needs to be striped off, same as W/F, SM and IR columns which contain the star symbol.
> 5. Input the correct datatypes for columns such as contract date, joined etc.
> 6. Height column records height of some players in cm and others in feet/inches.It should be converted to one standard measurement.
> 7. Weight column contains players weights in both kilogram and pounds,this should also be converted to one standard measurement.
> 8. Wage, Release clause and Value columns contains euro sign which should be removed, it also lists the ammount in M (for million)and K (for thousand). This would be removed and the subsequent values multiplied by 1000000 and 1000 respectively.
> 9. Contract column has date for players on active contract, and string (free and on loan) for those who are not.
> 10. Hits contains null/missing values which should be resolved.

<a id = 'Data_Cleaning'></a>
## Data Cleaning

In [12]:
# make a copy of our dataset
clean_fifa = fifa.copy()

Issue #1: Renaming some columns to make them readable

In [13]:
#renaming some columns 
clean_fifa.rename(columns ={'OVA':'Overall_analysis(%)','POT':'Players_potential(%)', 'Height': 'Height(m)', 'Weight':'Weight(kg)',
                       'BOV': 'Best_overal(%)', 'W/F': 'Weak_foot_rating', 'SM': 'Skill_moves_rating', 
                        'A/W': 'Attacking_work_rate', 'D/W': 'Defensive_work_rate', 'IR': 'Injury_rating', 'PAC': 'Pace',
                       'SHO': 'Shooting_attribute', 'PAS': 'Pass_accuracy', 'DRI': 'Dribbling_ability', 
                           'DEF': 'Defensive_ability', 'PHY': 'Physicality'},inplace=True)
clean_fifa.head(1)

ID      Name      LongName  \
0  158023  L. Messi  Lionel Messi   

                                           photoUrl  \
0  https://cdn.sofifa.com/players/158/023/21_60.png   

                                           playerUrl Nationality  Age  \
0  http://sofifa.com/player/158023/lionel-messi/2...   Argentina   33   

   Overall_analysis(%)  Players_potential(%)                  Club  \
0                   93                    93  \n\n\n\nFC Barcelona   

      Contract   Positions Height(m) Weight(kg) Preferred Foot  \
0  2004 ~ 2021  RW, ST, CF     170cm       72kg           Left   

   Best_overal(%) Best Position    Joined Loan Date End    Value   Wage  \
0              93            RW  1-Jul-04           NaN  €103.5M  €560K   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0        €138.4M        429        85         95                70   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   

   Goalkeeping  GK Diving  GK Handling  GK Kicking  GK Positioning  \
0           54          6           11          15              14   

   GK Reflexes  Total Stats  Base Stats Weak_foot_rating Skill_moves_rating  \
0            8         2231         466              4 ★                 4★   

  Attacking_work_rate Defensive_work_rate Injury_rating  Pace  \
0              Medium                 Low           5 ★    85   

   Shooting_attribute  Pass_accuracy  Dribbling_ability  Defensive_ability  \
0                  92             91                 95                 38   

   Physicality Hits  
0           65  771

Issue #2: Dropping some columns not needed for analysis

In [14]:
# dropping the columns not needed for analysis using their index and the pandas drop method
clean_fifa.drop(clean_fifa.columns[[2, 3, 4, 11]], axis = 1, inplace = True)

Issue #3: Stripping the unnecessary '\n\n\n\n' characters in club column and also stripping the star characters in the Injury rating, Weak_foot_rating and Skill_moves_rating columns

In [15]:
#stripping the unnecessary characters in columns such as club
clean_fifa['Club'] = clean_fifa['Club'].str.replace('\n\n\n\n','')
clean_fifa.head()

ID               Name Nationality  Age  Overall_analysis(%)  \
0  158023           L. Messi   Argentina   33                   93   
1   20801  Cristiano Ronaldo    Portugal   35                   92   
2  200389           J. Oblak    Slovenia   27                   91   
3  192985       K. De Bruyne     Belgium   29                   91   
4  190871          Neymar Jr      Brazil   28                   91   

   Players_potential(%)                 Club     Contract Height(m)  \
0                    93         FC Barcelona  2004 ~ 2021     170cm   
1                    92             Juventus  2018 ~ 2022     187cm   
2                    93      Atlético Madrid  2014 ~ 2023     188cm   
3                    91      Manchester City  2015 ~ 2023     181cm   
4                    91  Paris Saint-Germain  2017 ~ 2022     175cm   

  Weight(kg) Preferred Foot  Best_overal(%) Best Position     Joined  \
0       72kg           Left              93            RW   1-Jul-04   
1       83kg          Right              92            ST  10-Jul-18   
2       87kg          Right              91            GK  16-Jul-14   
3       70kg          Right              91           CAM  30-Aug-15   
4       68kg          Right              91            LW   3-Aug-17   

  Loan Date End    Value   Wage Release Clause  Attacking  Crossing  \
0           NaN  €103.5M  €560K        €138.4M        429        85   
1           NaN     €63M  €220K         €75.9M        437        84   
2           NaN    €120M  €125K        €159.4M         95        13   
3           NaN    €129M  €370K          €161M        407        94   
4           NaN    €132M  €270K        €166.5M        408        85   

   Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  \
0         95                70             91       88    470         96   
1         95                90             82       86    414         88   
2         11                15             43       13    109         12   
3         82                55             94       82    441         88   
4         87                62             87       87    448         95   

   Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0     93           94            91            96       451            91   
1     81           76            77            92       431            87   
2     13           14            40            30       307            43   
3     85           83            93            92       398            77   
4     88           89            81            95       453            94   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            80       91         94       95    389          86       68   
1            91       87         95       71    444          94       95   
2            60       67         88       49    268          59       78   
3            76       78         91       76    408          91       63   
4            89       96         91       83    357          80       62   

   Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
0       72        69          94        347          44             40   
1       84        78          93        353          63             29   
2       41        78          12        140          34             19   
3       89        74          91        408          76             66   
4       81        50          84        356          51             36   

   Positioning  Vision  Penalties  Composure  Defending  Marking  \
0           93      95         75         96         91       32   
1           95      82         84         95         84       28   
2           11      65         11         68         57       27   
3           88      94         84         91        186       68   
4           87      90         92         93         94       35   

   Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handl

In [16]:
#stripping the star character in columns Weak_foot_rating, Skill_moves_rating and Injury_rating
clean_fifa[['Weak_foot_rating','Skill_moves_rating','Injury_rating']] = clean_fifa[['Weak_foot_rating','Skill_moves_rating','Injury_rating']].apply(lambda x: x.str.replace('★',''))
clean_fifa.head(1)

ID      Name Nationality  Age  Overall_analysis(%)  \
0  158023  L. Messi   Argentina   33                   93   

   Players_potential(%)          Club     Contract Height(m) Weight(kg)  \
0                    93  FC Barcelona  2004 ~ 2021     170cm       72kg   

  Preferred Foot  Best_overal(%) Best Position    Joined Loan Date End  \
0           Left              93            RW  1-Jul-04           NaN   

     Value   Wage Release Clause  Attacking  Crossing  Finishing  \
0  €103.5M  €560K        €138.4M        429        85         95   

   Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  Curve  \
0                70             91       88    470         96     93   

   FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0           94            91            96       451            91   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            80       91         94       95    389          86       68   

   Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
0       72        69          94        347          44             40   

   Positioning  Vision  Penalties  Composure  Defending  Marking  \
0           93      95         75         96         91       32   

   Standing Tackle  Sliding Tackle  Goalkeeping  GK Diving  GK Handling  \
0               35              24           54          6           11   

   GK Kicking  GK Positioning  GK Reflexes  Total Stats  Base Stats  \
0          15              14            8         2231         466   

  Weak_foot_rating Skill_moves_rating Attacking_work_rate Defensive_work_rate  \
0               4                   4              Medium                 Low   

  Injury_rating  Pace  Shooting_attribute  Pass_accuracy  Dribbling_ability  \
0            5     85                  92             91                 95   

   Defensive_ability  Physicality Hits  
0                 38           65  771

Issue #4: Convert the player height column to one standard measuremnt, meters (m) 

In [17]:
# create a function to convert height to metres
def convert_to_metres(Height):
    if'cm' in Height:
        return float(Height.replace('cm', '')) / 100
    else:
        feet, inches = Height.split('\'')
        return round(float(feet) * 0.3048, 2) + round(float(inches.replace('"', '')) * 0.0254, 2)
    
    
# applying the function to the height column
clean_fifa['Height(m)'] = clean_fifa['Height(m)'].apply(convert_to_metres)

# Check the column 
clean_fifa['Height(m)'].unique()

array([1.7 , 1.87, 1.88, 1.81, 1.75, 1.84, 1.91, 1.78, 1.93, 1.85, 1.99,
       1.73, 1.68, 1.76, 1.77, 1.83, 1.8 , 1.89, 1.79, 1.95, 1.72, 1.82,
       1.86, 1.92, 1.65, 1.94, 1.67, 1.96, 1.63, 1.9 , 1.74, 1.69, 1.71,
       1.97, 2.  , 1.66, 1.88, 1.64, 1.98, 1.91, 1.93, 1.62, 2.01, 1.58,
       1.61, 1.6 , 2.03, 1.57, 1.56, 2.02, 1.59, 2.06, 1.55])

Issue #5: Convert the player weight column to one standard measuremnt, kilogram (kg).

In [18]:
# create a function to convert weight to kilogram
def convert_to_kilogram(weight):
    if'kg' in weight:
        return (weight.replace('kg', '')) 
    else:
        return round(float(weight.replace('lbs', '')) /2.20462, 2)
    
    
#applying the function to the height column
clean_fifa['Weight(kg)'] = clean_fifa['Weight(kg)'].apply(convert_to_kilogram)

# Check the column 
clean_fifa['Weight(kg)'].unique()

array(['72', '83', '87', '70', '68', '80', '71', '91', '73', '85', '92',
       '69', '84', '96', '81', '82', '75', '86', '89', '74', '76', '64',
       '78', '90', '66', '60', '94', '79', '67', '65', '59', '61', '93',
       '88', '97', '77', '62', '63', '95', '100', '58', 83.01, 81.19,
       78.02, 88.9, 79.83, 83.91, 77.11, 92.08, 76.2, 73.03, 66.22, 58.97,
       86.18, 78.93, 67.13, 74.84, 72.12, 87.09, 82.1, 63.05, 69.85,
       71.21, 73.94, '98', '103', '99', '102', '56', '101', '57', '55',
       '104', '107', '110', '53', '50', '54', '52'], dtype=object)

Issue #6: Remove the euro signs in Wage, Value and Release_clause columns, and remove the 'K' and 'M' and converting it to their values in thousands and millions respectively.

In [19]:
# define a function to remove '€' and convert the 'K' and 'M' to actual numbers
def convert_value(value_str):
    value = value_str.replace('€', '')
    if value_str.endswith ('K'):
        value = float(value[:-1]) * 1000
    elif value_str.endswith ('M'):
        value = float(value[:-1]) * 1000000
    else:
        value = float(value)
    if value.is_integer():
        return '{:.0f}'.format(value)
    else:
        return '{:.2f}'.format(value)

In [20]:
# apply the function to the wage, value and release clause columns
clean_fifa[['Value', 'Wage', 'Release Clause']] = clean_fifa[['Value', 'Wage', 'Release Clause']].applymap(convert_value)

In [21]:
clean_fifa.sample(10)

ID         Name Nationality  Age  Overall_analysis(%)  \
8883   241765      C. Akas     Nigeria   26                   66   
11247  255591  K. Nakayama       Japan   23                   64   
5371   189945     Y. Ravet      France   30                   70   
13341  246874     R. Gotts     England   20                   62   
7544   244258   D. Zagorac     Croatia   33                   67   
7636   229705   J. Duverne      France   22                   67   
14274  237930   S. Hausner     Denmark   20                   61   
16407  254085  F. Casanova     Uruguay   23                   58   
2118   211567     Gilberto      Brazil   27                   74   
4137   200818  Bruno Uvini      Brazil   29                   71   

       Players_potential(%)                   Club     Contract  Height(m)  \
8883                     69          Os Belenenses  2019 ~ 2021       1.88   
11247                    69            Yokohama FC  2019 ~ 2021       1.77   
5371                     70       Grenoble Foot 38  2020 ~ 2023       1.77   
13341                    77           Leeds United  2017 ~ 2022       1.70   
7544                     67          Dinamo Zagreb  2016 ~ 2022       1.86   
7636                     74      Stade Brestois 29  2019 ~ 2023       1.84   
14274                    76              Aarhus GF  2017 ~ 2023       1.87   
16407                    64  Centro Atlético Fénix  2019 ~ 2020       1.93   
2118                     75             SL Benfica  2020 ~ 2025       1.79   
4137                     72             Al Ittihad  2020 ~ 2022       1.87   

      Weight(kg) Preferred Foot  Best_overal(%) Best Position     Joined  \
8883          90           Left              66           LWB   8-Aug-19   
11247         65          Right              65           CAM   1-Jan-19   
5371          72          Right              70            RM   1-Jul-20   
13341         70          Right              66            RM   1-Jul-17   
7544          80          Right              67            GK  11-Jul-16   
7636          71          Right              69            CB   2-Sep-19   
14274         79           Left              63            CB  27-Feb-17   
16407         80           Left              58            GK   1-Oct-19   
2118          74          Right              74            RB  10-Aug-20   
4137          85           Left              72            CB  13-Jan-20   

      Loan Date End    Value   Wage Release Clause  Attacking  Crossing  \
8883            NaN  1000000   2000        1500000        242        64   
11247           NaN   875000   1000         795000        269        63   
5371            NaN  1500000   5000        3000000        315        75   
13341           NaN   950000  11000        1700000        268        55   
7544            NaN   325000    500         880000         60        11   
7636            NaN  1900000   5000        2100000        236        47   
14274           NaN   750000   2000         984000        195        35   
16407           NaN   250000    500         306000         65        12   
2118            NaN  4400000  11000       11400000        300        74   
4137            NaN  1700000  22000        3900000        216        19   

       Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  \
8883          34                49             61       34    254         62   
11247         55                37             61       53    299         65   
5371          59                50             73       58    342         72   
13341         56                47             69       41    268         62   
7544          10                11             19        9     88         18   
7636          25                67             65       32    247         55   
14274         20                60             57       23    200         48   
16407          7                14             27        5     77          8   
2118          52    

Issue #7: Resolve the contract column by creating a new column called 'Contract_status' which shows players on active contract, on loan and free agents. Two new columns would be created for contract start year and contract end year for start and end of contract respectively.

In [22]:
# split the contract column and create separate Contract_start_year and Contract_end_year
clean_fifa[['Contract_start_year', 'Contract_end_year']] = clean_fifa['Contract'].str.split('~', expand=True)

#display the result
clean_fifa.head()

ID               Name Nationality  Age  Overall_analysis(%)  \
0  158023           L. Messi   Argentina   33                   93   
1   20801  Cristiano Ronaldo    Portugal   35                   92   
2  200389           J. Oblak    Slovenia   27                   91   
3  192985       K. De Bruyne     Belgium   29                   91   
4  190871          Neymar Jr      Brazil   28                   91   

   Players_potential(%)                 Club     Contract  Height(m)  \
0                    93         FC Barcelona  2004 ~ 2021       1.70   
1                    92             Juventus  2018 ~ 2022       1.87   
2                    93      Atlético Madrid  2014 ~ 2023       1.88   
3                    91      Manchester City  2015 ~ 2023       1.81   
4                    91  Paris Saint-Germain  2017 ~ 2022       1.75   

  Weight(kg) Preferred Foot  Best_overal(%) Best Position     Joined  \
0         72           Left              93            RW   1-Jul-04   
1         83          Right              92            ST  10-Jul-18   
2         87          Right              91            GK  16-Jul-14   
3         70          Right              91           CAM  30-Aug-15   
4         68          Right              91            LW   3-Aug-17   

  Loan Date End      Value    Wage Release Clause  Attacking  Crossing  \
0           NaN  103500000  560000      138400000        429        85   
1           NaN   63000000  220000       75900000        437        84   
2           NaN  120000000  125000      159400000         95        13   
3           NaN  129000000  370000      161000000        407        94   
4           NaN  132000000  270000      166500000        408        85   

   Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  \
0         95                70             91       88    470         96   
1         95                90             82       86    414         88   
2         11                15             43       13    109         12   
3         82                55             94       82    441         88   
4         87                62             87       87    448         95   

   Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0     93           94            91            96       451            91   
1     81           76            77            92       431            87   
2     13           14            40            30       307            43   
3     85           83            93            92       398            77   
4     88           89            81            95       453            94   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            80       91         94       95    389          86       68   
1            91       87         95       71    444          94       95   
2            60       67         88       49    268          59       78   
3            76       78         91       76    408          91       63   
4            89       96         91       83    357          80       62   

   Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
0       72        69          94        347          44             40   
1       84        78          93        353          63             29   
2       41        78          12        140          34             19   
3       89        74          91        408          76             66   
4       81        50          84        356          51             36   

   Positioning  Vision  Penalties  Composure  Defending  Marking  \
0           93      95         75         96         91       32   
1           95      82         84         95         84       28   
2           11      65         11         68         57       27   
3           88      94         84         91        186       68   
4           87      90         92         93         94       35   

   Standing Tackle  Sliding Tackle  Goalkeep

In [23]:
# To categorize players on active contract, loan or free agents
conditions = [
    clean_fifa['Contract'] == 'Free',
     clean_fifa['Contract'].str.contains('On Loan')
]
values = ['Free agent', 'On Loan']

# use numpy .select() to create a new column based on conditions and values
clean_fifa['Contract_status'] = np.select(conditions, values, default = 'Active')

# display the result
clean_fifa.head()

ID               Name Nationality  Age  Overall_analysis(%)  \
0  158023           L. Messi   Argentina   33                   93   
1   20801  Cristiano Ronaldo    Portugal   35                   92   
2  200389           J. Oblak    Slovenia   27                   91   
3  192985       K. De Bruyne     Belgium   29                   91   
4  190871          Neymar Jr      Brazil   28                   91   

   Players_potential(%)                 Club     Contract  Height(m)  \
0                    93         FC Barcelona  2004 ~ 2021       1.70   
1                    92             Juventus  2018 ~ 2022       1.87   
2                    93      Atlético Madrid  2014 ~ 2023       1.88   
3                    91      Manchester City  2015 ~ 2023       1.81   
4                    91  Paris Saint-Germain  2017 ~ 2022       1.75   

  Weight(kg) Preferred Foot  Best_overal(%) Best Position     Joined  \
0         72           Left              93            RW   1-Jul-04   
1         83          Right              92            ST  10-Jul-18   
2         87          Right              91            GK  16-Jul-14   
3         70          Right              91           CAM  30-Aug-15   
4         68          Right              91            LW   3-Aug-17   

  Loan Date End      Value    Wage Release Clause  Attacking  Crossing  \
0           NaN  103500000  560000      138400000        429        85   
1           NaN   63000000  220000       75900000        437        84   
2           NaN  120000000  125000      159400000         95        13   
3           NaN  129000000  370000      161000000        407        94   
4           NaN  132000000  270000      166500000        408        85   

   Finishing  Heading Accuracy  Short Passing  Volleys  Skill  Dribbling  \
0         95                70             91       88    470         96   
1         95                90             82       86    414         88   
2         11                15             43       13    109         12   
3         82                55             94       82    441         88   
4         87                62             87       87    448         95   

   Curve  FK Accuracy  Long Passing  Ball Control  Movement  Acceleration  \
0     93           94            91            96       451            91   
1     81           76            77            92       431            87   
2     13           14            40            30       307            43   
3     85           83            93            92       398            77   
4     88           89            81            95       453            94   

   Sprint Speed  Agility  Reactions  Balance  Power  Shot Power  Jumping  \
0            80       91         94       95    389          86       68   
1            91       87         95       71    444          94       95   
2            60       67         88       49    268          59       78   
3            76       78         91       76    408          91       63   
4            89       96         91       83    357          80       62   

   Stamina  Strength  Long Shots  Mentality  Aggression  Interceptions  \
0       72        69          94        347          44             40   
1       84        78          93        353          63             29   
2       41        78          12        140          34             19   
3       89        74          91        408          76             66   
4       81        50          84        356          51             36   

   Positioning  Vision  Penalties  Composure  Defending  Marking  \
0           93      95         75         96         91       32   
1           95      82         84         95         84       28   
2           11      65         11         68         57       27   
3           88      94         84         91        186       68   
4           87      90         92         93         94       35   

   Standing Tackle  Sliding Tackle  Goalkeep

In [24]:
clean_fifa['Contract_status'].unique()

array(['Active', 'On Loan', 'Free agent'], dtype=object)

In [25]:
# Using the year in the contract column as contract_end_year for players on loan
clean_fifa.loc[clean_fifa['Contract_status'] == 'On Loan', 'Contract_end_year'] = clean_fifa['Contract'].apply(lambda x: x.split(',')[-1].strip()[:4])

In [26]:
# dropping the contract, loan date end and contract_start_year columns
clean_fifa.drop(['Contract', 'Loan Date End', 'Contract_start_year'], axis=1, inplace = True)
clean_fifa.head()

ID               Name Nationality  Age  Overall_analysis(%)  \
0  158023           L. Messi   Argentina   33                   93   
1   20801  Cristiano Ronaldo    Portugal   35                   92   
2  200389           J. Oblak    Slovenia   27                   91   
3  192985       K. De Bruyne     Belgium   29                   91   
4  190871          Neymar Jr      Brazil   28                   91   

   Players_potential(%)                 Club  Height(m) Weight(kg)  \
0                    93         FC Barcelona       1.70         72   
1                    92             Juventus       1.87         83   
2                    93      Atlético Madrid       1.88         87   
3                    91      Manchester City       1.81         70   
4                    91  Paris Saint-Germain       1.75         68   

  Preferred Foot  Best_overal(%) Best Position     Joined      Value    Wage  \
0           Left              93            RW   1-Jul-04  103500000  560000   
1          Right              92            ST  10-Jul-18   63000000  220000   
2          Right              91            GK  16-Jul-14  120000000  125000   
3          Right              91           CAM  30-Aug-15  129000000  370000   
4          Right              91            LW   3-Aug-17  132000000  270000   

  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0      138400000        429        85         95                70   
1       75900000        437        84         95                90   
2      159400000         95        13         11                15   
3      161000000        407        94         82                55   
4      166500000        408        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88    470         96     93           94            91   
1             82       86    414         88     81           76            77   
2             43       13    109         12     13           14            40   
3             94       82    441         88     85           83            93   
4             87       87    448         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96       451            91            80       91         94   
1            92       431            87            91       87         95   
2            30       307            43            60       67         88   
3            92       398            77            76       78         91   
4            95       453            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95    389          86       68       72        69          94   
1       71    444          94       95       84        78          93   
2       49    268          59       78       41        78          12   
3       76    408          91       63       89        74          91   
4       83    357          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0        347          44             40           93      95         75   
1        353          63             29           95      82         84   
2        140          34             19           11      65         11   
3        408          76             66           88      94         84   
4        356          51             36           87      90         92   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96         91       32               35              24   
1         95         84       28               32              24   
2         68         57       27               12              18   
3         91        186       68               65              53   
4         93         94       35               30              29   

   Goalk

Issue #8: Resolve the null/missing values and figures that contain 'K' in the hits column.

> Since hits represents the number of times a player's profile has been viewed on the sofia website, the 'nan' values can be replaced with 0 which means some players have had no views on the website. Figures that have k such as 2k can be multiplied by '1000'.

In [27]:
# creating a function to convert hits
def convert_hits(hit):
    if pd.isna(hit):
        return 0
    elif isinstance(hit, int):
        return hit
    elif 'K' in hit:
        return int(float(hit.replace('K', '')) * 1000)
    else:
        return int(float(hit))

# applying the function the Hits column
clean_fifa['Hits'] = clean_fifa['Hits'].apply(convert_hits)

Issue #9: Input the correct datatype for the columns in the dataset

In [28]:
# converting Weight and  Wage to integer
clean_fifa[['Weight(kg)', 'Wage']] = clean_fifa[['Weight(kg)', 'Wage']].astype(int) 

# converting Value and Release_clause to float
clean_fifa[['Release Clause', 'Value']] = clean_fifa[['Release Clause', 'Value']].astype(float)

# Converting ID column to string
clean_fifa['ID'] = clean_fifa['ID'].astype(str)

# converting Joined and Contract_end_year to datetime
clean_fifa['Joined'] = pd.to_datetime(clean_fifa['Joined'])

# converting contract_status to category
clean_fifa[['Contract_status', 'Skill_moves_rating']] = clean_fifa[['Contract_status', 'Skill_moves_rating']].astype('category')

In [29]:
# Converting Weak_foot_rating, Skill_moves_rating, Injury_rating, Attacking_work_rate and Defensive_work_rate to ordered category
ordinal_dict = {'Attacking_work_rate': ['Low', 'Medium', 'High'], 'Defensive_work_rate': ['Low', 'Medium', 'High'],
               'Weak_foot_rating': ['1 ', '2 ', '3 ', '4 ', '5 '], 'Injury_rating': ['1 ', '2 ', '3 ', '4 ', '5 ']}
for key, value in ordinal_dict.items():
    order = pd.api.types.CategoricalDtype(categories = value, ordered = True)
    clean_fifa[key] = clean_fifa[key].astype(order)

In [30]:
# confirm the datatype for each column
clean_fifa.dtypes

ID                              object
Name                            object
Nationality                     object
Age                              int64
Overall_analysis(%)              int64
Players_potential(%)             int64
Club                            object
Height(m)                      float64
Weight(kg)                       int32
Preferred Foot                  object
Best_overal(%)                   int64
Best Position                   object
Joined                  datetime64[ns]
Value                          float64
Wage                             int32
Release Clause                 float64
Attacking                        int64
Crossing                         int64
Finishing                        int64
Heading Accuracy                 int64
Short Passing                    int64
Volleys                          int64
Skill                            int64
Dribbling                        int64
Curve                            int64
FK Accuracy              

In [31]:
# summing up each stats
clean_fifa["Attacking"] = fifa["Crossing"] + fifa["Heading Accuracy"] + fifa["Finishing"] + fifa["Short Passing"] + fifa["Volleys"]
clean_fifa["Skill"] = fifa["Dribbling"] + fifa["FK Accuracy"] + fifa["Curve"] + fifa["Long Passing"] + fifa["Ball Control"]
clean_fifa["Power"] = fifa["Shot Power"] + fifa["Jumping"] + fifa["Stamina"] + fifa["Strength"] + fifa["Long Shots"]
clean_fifa["Defending"] = fifa["Standing Tackle"] + fifa["Sliding Tackle"] + fifa["Marking"]
clean_fifa["Mentality"] = fifa["Aggression"] + fifa["Interceptions"] + fifa["Positioning"] + fifa["Vision"] + fifa["Penalties"] + fifa["Composure"]
clean_fifa["Goalkeeping"] = fifa["GK Diving"] + fifa["GK Handling"] + fifa["GK Kicking"] + fifa["GK Positioning"] + fifa["GK Reflexes"]
clean_fifa["Movement"] = fifa["Acceleration"] + fifa["Sprint Speed"] + fifa["Agility"] + fifa["Reactions"] + fifa["Balance"]

In [32]:
# rounding each of the main skills / 100
clean_fifa["Attacking"] = round(fifa["Attacking"]/5)
clean_fifa["Skill"] = round(fifa["Skill"]/5)
clean_fifa["Movement"] = round(fifa["Movement"]/5)
clean_fifa["Power"] = round(fifa["Power"]/5)
clean_fifa["Mentality"] = round(fifa["Mentality"]/6)
clean_fifa["Defending"] = round(fifa["Defending"]/3)
clean_fifa["Goalkeeping"] = round(fifa["Goalkeeping"]/5)

In [33]:
clean_fifa.head()

ID               Name Nationality  Age  Overall_analysis(%)  \
0  158023           L. Messi   Argentina   33                   93   
1   20801  Cristiano Ronaldo    Portugal   35                   92   
2  200389           J. Oblak    Slovenia   27                   91   
3  192985       K. De Bruyne     Belgium   29                   91   
4  190871          Neymar Jr      Brazil   28                   91   

   Players_potential(%)                 Club  Height(m)  Weight(kg)  \
0                    93         FC Barcelona       1.70          72   
1                    92             Juventus       1.87          83   
2                    93      Atlético Madrid       1.88          87   
3                    91      Manchester City       1.81          70   
4                    91  Paris Saint-Germain       1.75          68   

  Preferred Foot  Best_overal(%) Best Position     Joined        Value  \
0           Left              93            RW 2004-07-01  103500000.0   
1          Right              92            ST 2018-07-10   63000000.0   
2          Right              91            GK 2014-07-16  120000000.0   
3          Right              91           CAM 2015-08-30  129000000.0   
4          Right              91            LW 2017-08-03  132000000.0   

     Wage  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  560000     138400000.0       86.0        85         95                70   
1  220000      75900000.0       87.0        84         95                90   
2  125000     159400000.0       19.0        13         11                15   
3  370000     161000000.0       81.0        94         82                55   
4  270000     166500000.0       82.0        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88   94.0         96     93           94            91   
1             82       86   83.0         88     81           76            77   
2             43       13   22.0         12     13           14            40   
3             94       82   88.0         88     85           83            93   
4             87       87   90.0         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96      90.0            91            80       91         94   
1            92      86.0            87            91       87         95   
2            30      61.0            43            60       67         88   
3            92      80.0            77            76       78         91   
4            95      91.0            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95   78.0          86       68       72        69          94   
1       71   89.0          94       95       84        78          93   
2       49   54.0          59       78       41        78          12   
3       76   82.0          91       63       89        74          91   
4       83   71.0          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       58.0          44             40           93      95         75   
1       59.0          63             29           95      82         84   
2       23.0          34             19           11      65         11   
3       68.0          76             66           88      94         84   
4       59.0          51             36           87      90         92   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96       30.0       32               35              24   
1         95       28.0       28               32              24   
2         68       19.0       27               12              18   
3         91       62.0       68               65              53   
4         93       31.0       35               30     

#### Creating a new column known as Body_mass_index (BMI) which is an important factor to evaluate a footballer's health and fitness. Formular for BMI is weight(kg) / (height(m))^2

In [34]:
# create a BmI column using the formular
clean_fifa['Body_mass_index'] = clean_fifa['Weight(kg)'] / (clean_fifa['Height(m)'] ** 2)

# round the BMI column to two decimal places
clean_fifa['Body_mass_index'] = clean_fifa['Body_mass_index'].round(2)

In [35]:
clean_fifa.head()

ID               Name Nationality  Age  Overall_analysis(%)  \
0  158023           L. Messi   Argentina   33                   93   
1   20801  Cristiano Ronaldo    Portugal   35                   92   
2  200389           J. Oblak    Slovenia   27                   91   
3  192985       K. De Bruyne     Belgium   29                   91   
4  190871          Neymar Jr      Brazil   28                   91   

   Players_potential(%)                 Club  Height(m)  Weight(kg)  \
0                    93         FC Barcelona       1.70          72   
1                    92             Juventus       1.87          83   
2                    93      Atlético Madrid       1.88          87   
3                    91      Manchester City       1.81          70   
4                    91  Paris Saint-Germain       1.75          68   

  Preferred Foot  Best_overal(%) Best Position     Joined        Value  \
0           Left              93            RW 2004-07-01  103500000.0   
1          Right              92            ST 2018-07-10   63000000.0   
2          Right              91            GK 2014-07-16  120000000.0   
3          Right              91           CAM 2015-08-30  129000000.0   
4          Right              91            LW 2017-08-03  132000000.0   

     Wage  Release Clause  Attacking  Crossing  Finishing  Heading Accuracy  \
0  560000     138400000.0       86.0        85         95                70   
1  220000      75900000.0       87.0        84         95                90   
2  125000     159400000.0       19.0        13         11                15   
3  370000     161000000.0       81.0        94         82                55   
4  270000     166500000.0       82.0        85         87                62   

   Short Passing  Volleys  Skill  Dribbling  Curve  FK Accuracy  Long Passing  \
0             91       88   94.0         96     93           94            91   
1             82       86   83.0         88     81           76            77   
2             43       13   22.0         12     13           14            40   
3             94       82   88.0         88     85           83            93   
4             87       87   90.0         95     88           89            81   

   Ball Control  Movement  Acceleration  Sprint Speed  Agility  Reactions  \
0            96      90.0            91            80       91         94   
1            92      86.0            87            91       87         95   
2            30      61.0            43            60       67         88   
3            92      80.0            77            76       78         91   
4            95      91.0            94            89       96         91   

   Balance  Power  Shot Power  Jumping  Stamina  Strength  Long Shots  \
0       95   78.0          86       68       72        69          94   
1       71   89.0          94       95       84        78          93   
2       49   54.0          59       78       41        78          12   
3       76   82.0          91       63       89        74          91   
4       83   71.0          80       62       81        50          84   

   Mentality  Aggression  Interceptions  Positioning  Vision  Penalties  \
0       58.0          44             40           93      95         75   
1       59.0          63             29           95      82         84   
2       23.0          34             19           11      65         11   
3       68.0          76             66           88      94         84   
4       59.0          51             36           87      90         92   

   Composure  Defending  Marking  Standing Tackle  Sliding Tackle  \
0         96       30.0       32               35              24   
1         95       28.0       28               32              24   
2         68       19.0       27               12              18   
3         91       62.0       68               65              53   
4         93       31.0       35               30     

In [36]:
# save the cleaned dataset 
clean_fifa.to_csv('./Clean_fifa_2021.csv')